go-controller/pkg/ovn/obj_retry.go

```diff
@@ -126,18 +189,1067 @@ func (r *retryObjs) getObjRetryEntry(key string) *retryObjEntry {
+
+	case factory.NodeType:
+		name = "syncNodes"
+		syncRetriableFunc = oc.syncNodesRetriable
+
+	case factory.PeerServiceTypeVar,
+		factory.PeerNamespaceAndPodSelectorTypeVar:
+		name = ""
+		syncRetriableFunc = nil
+
+	case factory.PeerPodSelectorTypeVar:
+		name = "PeerPodSelector"
+		extraParameters := r.extraParameters.(*NetworkPolicyExtraParameters)
+		syncRetriableFunc = func(objs []interface{}) error {
+			return oc.handlePeerPodSelectorAddUpdate(extraParameters.gp, objs...)
+		}
+
+	case factory.PeerPodForNamespaceAndPodSelectorType:
+		name = "PeerPodForNamespaceAndPodSelector"
+		extraParameters := r.extraParameters.(*NetworkPolicyExtraParameters)
+		syncRetriableFunc = func(objs []interface{}) error {
+			return oc.handlePeerPodSelectorAddUpdate(extraParameters.gp, objs...)
+		}
+
+	case factory.PeerNamespaceSelectorTypeVar:
+		name = "PeerNamespaceSelector"
+		extraParameters := r.extraParameters.(*NetworkPolicyExtraParameters)
+		// the function below will never fail, so there's no point in making it retriable...
+		syncFunc = func(i []interface{}) {
+			// This needs to be a write lock because there's no locking around 'gress policies
+			extraParameters.np.Lock()
+			defer extraParameters.np.Unlock()
+			// We load the existing address set into the 'gress policy.
+			// Notice that this will make the AddFunc for this initial
+			// address set a noop.
+			// The ACL must be set explicitly after setting up this handler
+			// for the address set to be considered.
+			extraParameters.gp.addNamespaceAddressSets(i)
+		}
+
+	case factory.LocalPodSelectorTypeVar:
+		name = "LocalPodSelectorTypeVar"
+		syncRetriableFunc = r.syncFunc
+
+	default:
+		return nil, fmt.Errorf("no sync function for object type %v", r.oType)
+	}
+
+	if syncFunc == nil {
+		syncFunc = func(objects []interface{}) {
+			if syncRetriableFunc == nil {
 				return
 			}
+			oc.syncWithRetry(name, func() error { return syncRetriableFunc(objects) })
 		}
-	}()
+	}
+	return syncFunc, nil
+}
+
+// Given an object and its type, isObjectInTerminalState returns true if the object is a in terminal state.
+// This is used now for pods that are either in a PodSucceeded or in a PodFailed state.
+func (oc *Controller) isObjectInTerminalState(objType reflect.Type, obj interface{}) bool {
+	switch objType {
+	case factory.PodType,
+		factory.PeerPodSelectorTypeVar,
+		factory.PeerPodForNamespaceAndPodSelectorType,
+		factory.LocalPodSelectorTypeVar:
+		pod := obj.(*kapi.Pod)
+		return util.PodCompleted(pod)
+
+	default:
+		return false
+	}
+}
+
+// processObjectInTerminalState is executed when an object has been added or updated and is actually in a terminal state
+// already. The add or update event is not valid for such object, which we now remove from the cluster in order to
+// free its resources. (for now, this applies to completed pods)
+func (oc *Controller) processObjectInTerminalState(objectsToRetry *retryObjs, obj interface{}, key, event string) {
```

Comment:
for event arg maybe use a custom type rather than passing an "add" or "update" string